src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -15,103 +15,129 @@ namespace System.Net.Http
+            }
 
-        protected internal override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) =>
-            SendAsyncCore(request, async: true, cancellationToken).AsTask();
+            activity = null;
 
-        private async ValueTask<HttpResponseMessage> SendAsyncCore(HttpRequestMessage request, bool async,
-            CancellationToken cancellationToken)
-        {
-            // HttpClientHandler is responsible to call static DiagnosticsHandler.IsEnabled() before forwarding request here.
-            // It will check if propagation is on (because parent Activity exists or there is a listener) or off (forcibly disabled)
-            // This code won't be called unless consumer unsubscribes from DiagnosticListener right after the check.
-            // So some requests happening right after subscription starts might not be instrumented. Similarly,
-            // when consumer unsubscribes, extra requests might be instrumented
+            bool currentActivitySet = Activity.Current is not null;
+            bool diagnosticListenerEnabled = s_diagnosticListener.IsEnabled();
 
-            if (request == null)
+            if (!currentActivitySet && !diagnosticListenerEnabled)
             {
-                throw new ArgumentNullException(nameof(request), SR.net_http_handler_norequest);
+                return false;
             }
 
-            Activity? activity = null;
-            DiagnosticListener diagnosticListener = Settings.s_diagnosticListener;
+            bool diagnosticListenerActivityEnabled = diagnosticListenerEnabled && s_diagnosticListener.IsEnabled(ActivityName, request);
 
-            // if there is no listener, but propagation is enabled (with previous IsEnabled() check)
-            // do not write any events just start/stop Activity and propagate Ids
-            if (!diagnosticListener.IsEnabled())
+            if (diagnosticListenerActivityEnabled || currentActivitySet)
             {
-                activity = new Activity(DiagnosticsHandlerLoggingStrings.ActivityName);
-                activity.Start();
-                InjectHeaders(activity, request);
+                activity = s_activitySource.CreateActivity(ActivityName, ActivityKind.Client);
 
-                try
+                if (activity is null && (diagnosticListenerActivityEnabled || !s_activitySource.HasListeners()))
                 {
-                    return async ?
-                        await base.SendAsync(request, cancellationToken).ConfigureAwait(false) :
-                        base.Send(request, cancellationToken);
-                }
-                finally
-                {
-                    activity.Stop();
+                    // ActivitySource decided not to create an Activity
+                    // If that is because there were no listeners, manually create an Activity to maintain the behaviour before ActivitySource was introduced
```

Comment:
I've pushed [an alternative to a separate branch](https://github.com/MihaZupan/runtime-1/blob/53c2c527128d0f8935ae42efa5f8d986dbc668d9/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L63-L84).

Looking at [what was done when ActivitySource was added to AspNet](https://github.com/dotnet/aspnetcore/blob/ab87c90db386a5c442e1299b8dbc24a2d37a3754/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L57), it looks like `ActivitySource` was added as another alternative for deciding enabled rather than just a sampling decision.

I suppose that is the behaviour you were expecting here as well?

Reply:
Yeah that is what I had anticipated : )